Lib/WindowsRuntime/Blob/BlobReadStream.cs

```diff
@@ -95,7 +95,7 @@ public override int Read(byte[] buffer, int offset, int count)
         /// <param name="count">The maximum number of bytes to read.</param>
         /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
         /// <returns>A task that represents the asynchronous read operation.</returns>
-        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
```

Comment:
is the slight change of behavior here by eliding the async keyword gaining us much perf improvement? other than skipping the internal state machine?

Reply:
@erezvani1529 there should be no change in behaviour, solely performance gain.
If all the method returns is a task, rather than awaiting for the method _and_ for the operation in the method, only a single await (state machine) is involved.

Reply:
I am referring to the change of behavior with Exceptions and where to expect them, when the method call and await are separated.

Reply:
Exception will occur where you await the task, if I understood your question correctly. As Task returned by a method is not executed until awaited.

Reply:
That is true if the async keyword exists or the method call and await aren't separated. Stephen Cleary has a great [blog-post](http://blog.stephencleary.com/2016/12/eliding-async-await.html) about this as well, which could be helpful.

Reply:
@SeanFeldman https://github.com/Azure/azure-storage-net/pull/579#issuecomment-347668081

Reply:
@erezvani1529 regarding your comment in https://github.com/Azure/azure-storage-net/pull/581#discussion_r155389552

Eliding the keyword on the hot path means you can save the extra allocations of the statemachine display class. .NET 2.1 will contain improvements that will optimize the statemachine slightly see https://github.com/dotnet/coreclr/pull/14178 and https://github.com/dotnet/coreclr/pull/13105 but still there is allocation overhead. 

See for example

``` ini

BenchmarkDotNet=v0.10.11, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Core i7-4980HQ CPU 2.80GHz (Haswell), ProcessorCount=8
Frequency=2728068 Hz, Resolution=366.5598 ns, Timer=TSC
  [Host]     : .NET Framework 4.7 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2115.0
  DefaultJob : .NET Framework 4.7 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2115.0


```
|              Method |     Mean |     Error |    StdDev |    StdErr |      Min |       Q1 |   Median |       Q3 |      Max |  Op/s | Scaled | Allocated |
|-------------------- |---------:|----------:|----------:|----------:|---------:|---------:|---------:|---------:|---------:|------:|-------:|----------:|
|    WithStatemachine | 15.61 ms | 0.0166 ms | 0.0155 ms | 0.0040 ms | 15.58 ms | 15.60 ms | 15.61 ms | 15.62 ms | 15.63 ms | 64.06 |   1.00 |     640 B |
| WithoutStatemachine | 15.61 ms | 0.0128 ms | 0.0120 ms | 0.0031 ms | 15.59 ms | 15.60 ms | 15.61 ms | 15.62 ms | 15.63 ms | 64.06 |   1.00 |     384 B |

``` ini

BenchmarkDotNet=v0.10.11, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Core i7-4980HQ CPU 2.80GHz (Haswell), ProcessorCount=8
Frequency=2728068 Hz, Resolution=366.5598 ns, Timer=TSC
  [Host]     : .NET Framework 4.7 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2115.0
  Job-YTNNJT : .NET Core 2.0.3 (Framework 4.6.25815.02), 64bit RyuJIT

Toolchain=.NET Core 2.0  

```
|              Method |     Mean |     Error |    StdDev |    StdErr |      Min |       Q1 |   Median |       Q3 |      Max |  Op/s | Scaled | Allocated |
|-------------------- |---------:|----------:|----------:|----------:|---------:|---------:|---------:|---------:|---------:|------:|-------:|----------:|
|    WithStatemachine | 15.61 ms | 0.0155 ms | 0.0138 ms | 0.0037 ms | 15.58 ms | 15.60 ms | 15.60 ms | 15.61 ms | 15.63 ms | 64.08 |   1.00 |     528 B |
| WithoutStatemachine | 15.60 ms | 0.0186 ms | 0.0174 ms | 0.0045 ms | 15.57 ms | 15.59 ms | 15.60 ms | 15.62 ms | 15.63 ms | 64.08 |   1.00 |     312 B |

As you can imagine the allocations here would add up pretty quickly. 

It would be possible to unbundle those things from the PR. but I do think it is worth keeping.